src/sage/rings/function_field/drinfeld_modules/morphism.py

```diff
@@ -404,3 +419,365 @@ def ore_polynomial(self):
+        l = len(us)
+        row = [KT([us[i][j] for i in range(l)]) for j in range(r)]
+        rows = [row]
+
+        # The next rows:
+        # each row is obtained from the previous one by
+        # applying the semi-linear transformation f |-> t*f
+        inv = K(phiT[r]).inverse()
+        B = inv * phiT
+        T = KT.gen()
+        for i in range(1, r):
+            row = [c.map_coefficients(Frob) for c in row]
+            row = [(inv*T - B[0]) * row[-1]] + [row[j-1] - B[j]*row[-1] for j in range(1, r)]
+            rows.append(row)
+
+        return matrix(KT, rows)
+
+    def norm(self, ideal=True):
+        r"""
+        Return the norm of this isogeny.
+
+        INPUT:
+
+        - ``ideal`` -- a boolean (default: ``True``); if ``True``,
+          return the norm as an ideal in the function ring of the Drinfeld
+          modules; if ``False``, return the norm as an element in this
+          function ring (only relevant for endomorphisms)
+
+        EXAMPLES::
+
+            sage: Fq = GF(5)
+            sage: A.<T> = Fq[]
+            sage: K.<z> = Fq.extension(3)
+            sage: phi = DrinfeldModule(A, [z, 0, 1, z])
+            sage: t = phi.ore_variable()
+            sage: f = phi.hom(t + 1)
+            sage: f.norm()
+            Principal ideal (T + 4) of Univariate Polynomial Ring in T over Finite Field of size 5
+
+        The norm of the Frobenius endomorphism is equal to the characteristic::
+
+            sage: F = phi.frobenius_endomorphism()
+            sage: F.norm()
+            Principal ideal (T^3 + 3*T + 3) of Univariate Polynomial Ring in T over Finite Field of size 5
+            sage: phi.characteristic()
+            T^3 + 3*T + 3
+
+        For `a` in the underlying function ring, the norm of the
+        endomorphism given by `\phi_a` is `a^r` where `r` is the rank::
+
+            sage: g = phi.hom(T)
+            sage: g.norm()
+            Principal ideal (T^3) of Univariate Polynomial Ring in T over Finite Field of size 5
+
+            sage: h = phi.hom(T+1)
+            sage: h.norm()
+            Principal ideal (T^3 + 3*T^2 + 3*T + 1) of Univariate Polynomial Ring in T over Finite Field of size 5
+
+        For endomorphisms, the norm is not an ideal of `A` but it makes
+        sense as an actual element of `A`. We can get this element by passing
+        in the argument ``ideal=False``::
+
+            sage: phi.hom(2*T).norm(ideal=False)
+            3*T^3
+
+            sage: f.norm(ideal=False)
+            Traceback (most recent call last):
+            ...
+            ValueError: norm is defined as an actual element only for endomorphisms
+
+        """
+        nu = self._motive_matrix().det()
+        # We cast to A
+        A = self.domain().function_ring()
+        if ideal:
+            nu = A([c.in_base() for c in nu.monic().list()])
+            return A.ideal(nu)
+        elif self.domain() is self.codomain():
+            nu = A([c.in_base() for c in nu.list()])
```

Comment:
If returning the variable without further processing, maybe remove the intermediate variable.